### PR TITLE
[ConstraintSolver] Make type matching more transparent in case of failures

### PR DESCRIPTION
**Pull request by xedin**
This is intended as a series of changes to make `matchTypes` and its specialized forms
to be more transparent in case it fails to match something. This is going to help with 
type-checker diagnostics since solver would be able to capture more information about 
each failure.

### CODE REVIEW

File: lib/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -2228,51 +2228,77 @@ class ConstraintSystem {
...
 
+  class TypeMatchResult {
+    SolutionKind Kind;
+
+  public:
+    inline bool isSuccess() const { return Kind == SolutionKind::Solved; }
+    inline bool isFailure() const { return Kind == SolutionKind::Error; }
+    inline bool isAmbiguous() const { return Kind == SolutionKind::Unsolved; }
+
+    static TypeMatchResult success(ConstraintSystem &cs) {
```
Comment (rudkx) [inline lib/Sema/ConstraintSystem.h]: You may want to consider making helpers on ConstraintSystem like:
```
  TypeMatchResult getTypeMatchSuccess() {
    return TypeMatchResult::success(*this);
  }
  TypeMatchResult getTypeMatchFailure(ConstraintLocatorBuilder *locator) {
    return TypeMatchResult::failure(*this, locator);
  }
```

That would make some the other changes simpler/cleaner.

**Comment by rudkx** [comment]
Otherwise this looks like a good start.

File: lib/Sema/ConstraintSystem.h (hunk 1 above)
Comment (xedin) [reply lib/Sema/ConstraintSystem.h]: Good point! Working on that right now.

File: lib/Sema/CSSimplify.cpp [hunk 2]
```diff
@@ -664,10 +666,10 @@ matchCallArguments(ConstraintSystem &cs, ConstraintKind kind,
       if (tupleArgType->getNumElements() != 1 ||
           (!cs.getASTContext().isSwiftVersion3() &&
            tupleArgType->getElement(0).hasName())) {
-        return ConstraintSystem::SolutionKind::Error;
+        return ConstraintSystem::TypeMatchResult::failure(cs, locator);
       }
     }
-    return ConstraintSystem::SolutionKind::Solved;
+    return ConstraintSystem::TypeMatchResult::success(cs);
```
Comment (rudkx) [inline lib/Sema/CSSimplify.cpp]: This could be cs.getTypeMatchFailure(), right?

Comment (rudkx) [reply lib/Sema/CSSimplify.cpp]: Same with some of the other places above and below.

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Hm, I thought I auto replaced everything, but looks like not :(

**Comment by xedin** [comment]
@swift-ci please clean test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
Going to merge these changes to `matchTypes` and have other changes to constraint solver in the separate PRs.

**Comment by rudkx** [comment]
SGTM!

**Pull request merged**